pulp_container/app/viewsets.py

```diff
@@ -937,26 +937,29 @@ def build_image(self, request, pk):
         )
 
         serializer.is_valid(raise_exception=True)
+        serialized_data = serializer.deferred_validation(serializer.validated_data)
 
-        containerfile = serializer.validated_data["containerfile_artifact"]
-        try:
-            containerfile.save()
-        except IntegrityError:
-            containerfile = Artifact.objects.get(sha256=containerfile.sha256)
-            containerfile.touch()
-        tag = serializer.validated_data["tag"]
+        containerfile_tempfile_pk = None
+        if containerfile := serialized_data.get("containerfile", None):
+            temp_file = PulpTemporaryFile(file=containerfile)
+            temp_file.save()
+            containerfile_tempfile_pk = str(temp_file.pk)
 
-        artifacts = serializer.validated_data["artifacts"]
-        Artifact.objects.filter(pk__in=artifacts.keys()).touch()
+        containerfile_artifact_pk = serialized_data.get("containerfile_artifact_pk", None)
+        content_artifact_pks = serialized_data.get("content_artifact_pks", None)
+        build_context_repo = serialized_data.get("build_context_repo", None)
+        tag = serialized_data["tag"]
 
         result = dispatch(
             tasks.build_image_from_containerfile,
             exclusive_resources=[repository],
+            shared_resources=[build_context_repo],
             kwargs={
-                "containerfile_pk": str(containerfile.pk),
+                "containerfile_artifact_pk": containerfile_artifact_pk,
+                "containerfile_tempfile_pk": containerfile_tempfile_pk,
                 "tag": tag,
                 "repository_pk": str(repository.pk),
-                "artifacts": artifacts,
+                "content_artifact_pks": content_artifact_pks,
```

Comment:
Oh, I was thinking we were going to pass the build context and the containerfile_name to the task.

Reply:
Maybe we should keep the old task untouched to be deprecated and removed in a few releases and design a new one with a clean interface next to it.